satellite/satellitedb/nodeselection.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/common/pb"
+	"storj.io/common/storj"
+	"storj.io/private/version"
+	"storj.io/storj/satellite/overlay"
+)
+
+func (cache *overlaycache) SelectStorageNodes(ctx context.Context, totalNeededNodes, newNodeCount int, criteria *overlay.NodeCriteria) (nodes []*overlay.SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+	if totalNeededNodes == 0 {
+		return nil, nil
+	}
+
+	if newNodeCount > totalNeededNodes {
+		return nil, Error.New("requested new node count can't exceed requested total node count")
+	}
+
+	needNewNodes := newNodeCount
+	needReputableNodes := totalNeededNodes - needNewNodes
+
+	receivedNewNodes := 0
+	receivedNodeNetworks := make(map[string]struct{})
+
+	var excludedIDs []storj.NodeID
+	excludedIDs = append(excludedIDs, criteria.ExcludedIDs...)
+
+	var excludedNetworks []string
+	excludedNetworks = append(excludedNetworks, criteria.ExcludedNetworks...)
+
+	for i := 0; i < 3; i++ {
+		reputableNodes, newNodes, err := cache.selectStorageNodesOnce(ctx, needReputableNodes, needNewNodes, criteria, excludedIDs, excludedNetworks)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, node := range newNodes {
+			// checking for last net collision among reputable and new nodes since we can't check within the query
+			if _, ok := receivedNodeNetworks[node.LastNet]; ok {
+				continue
+			}
+			excludedIDs = append(excludedIDs, node.ID)
+			excludedNetworks = append(excludedNetworks, node.LastNet)
+			nodes = append(nodes, node)
+			needNewNodes--
+			receivedNewNodes++
+
+			if criteria.DistinctIP {
+				receivedNodeNetworks[node.LastNet] = struct{}{}
+			}
+		}
+		for _, node := range reputableNodes {
+			if _, ok := receivedNodeNetworks[node.LastNet]; ok {
+				continue
+			}
+			excludedIDs = append(excludedIDs, node.ID)
+			excludedNetworks = append(excludedNetworks, node.LastNet)
+			nodes = append(nodes, node)
+			needReputableNodes--
+
+			if criteria.DistinctIP {
+				receivedNodeNetworks[node.LastNet] = struct{}{}
+			}
+		}
+
+		if needNewNodes > 0 && receivedNewNodes == 0 {
+			needReputableNodes += needNewNodes
+			needNewNodes = 0
+		}
+
+		if needReputableNodes <= 0 && needNewNodes <= 0 {
+			break
+		}
+	}
+	return nodes, nil
```

Comment:
nit - should there be a final check to test if len(nodes) is at least the amount requested, and return an error otherwise? Or is the behavior we want to return the list of nodes, even if there are not enough?

Reply:
hmm I see in the existing code that we return the final amount without checking length and do a `if len(nodes) < reputableNodeCount` check in the overlay/service level to return an error from FindStorageNodesWithPreferences. I'm kind of fine with leaving that as is, unless you think it would be better to return the error from this function.

Reply:
oh yeah I see that it is happening in service.go. That is fine to me.